fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/temp/rules/FilterWithFieldWithComparisonRule.java

```diff
@@ -67,48 +63,36 @@ public ChangesMade onMatch(@Nonnull PlannerRuleCall call) {
         }
 
         final FieldWithComparison singleField = call.get(filterMatcher);
-        final Comparisons.Comparison comparison = call.get(comparisonMatcher);
-        final ScanComparisons scanComparisons = ScanComparisons.from(comparison);
-        if (scanComparisons == null) {
+        if (ScanComparisons.getComparisonType(singleField.getComparison()).equals(ScanComparisons.ComparisonType.NONE)) {
             // This comparison cannot be accomplished with a single scan.
             return ChangesMade.NO_CHANGE;
         }
 
-        FieldKeyExpression field = firstField(call.getContext().getCommonPrimaryKey());
-        if (field != null && Objects.equals(singleField.getFieldName(), field.getFieldName())) {
-            call.yield(SingleExpressionRef.of(new RecordQueryScanPlan(scanComparisons, false)));
-            return ChangesMade.MADE_CHANGES;
+
+        final KeyExpression commonPrimaryKey = call.getContext().getCommonPrimaryKey();
+        KeyExpressionComparisons comparisons;
+        Optional<KeyExpressionComparisons> matchedComparisons;
+        if (commonPrimaryKey != null) {
+            comparisons = new KeyExpressionComparisons(commonPrimaryKey);
+            matchedComparisons = comparisons.matchWith(singleField);
+            if (matchedComparisons.isPresent()) {
+                // TODO improve
```

Comment:
What's the improvement here that's envisioned?

Reply:
Basically the same thing as above, where we don't properly support primary key scans in all the rules; for example, there's no equivalent of the `LogicalIndexScanExpression` for `RecordQueryScanPlan`, and I don't really want to add one. The real solution is to treat primary key and index scans the same way for as long as possible which I've implemented in the next planner PR. 

I expanded the comment here in the meantime.